UCR - remove records from datasource where not pass filters

lwyszomi opened the pull request:
@emord 
http://manage.dimagi.com/default.asp?246557#1300659

lwyszomi committed:
UCR - remove records from datasource where not pass filters

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -137,10 +137,14 @@ def process_change(self, pillow_instance, change):
             ensure_matched_revisions(change)
+
+            def exist_in_database(document):
+                return table.get_query_object().filter_by(doc_id=document['_id']).count() > 0
```
should use `.exists()` rather than `.count()`

lwyszomi replied:
@snopoke updated

lwyszomi committed:
Change count to exists

emord reviewed (approved):
Seems ok. to me @lwyszomi would it be possible to write at test for this?

Going to hold off on merging it until our pillows are in a ok state on prod

lwyszomi committed:
added test

lwyszomi committed:
fix test

dimagimon commented on corehq/apps/userreports/pillow.py:
```diff
@@ -5,6 +5,8 @@
 
 from alembic.autogenerate.api import compare_metadata
 import six
+from sqlalchemy import select
```
F401 'select' imported but unused

dimagimon commented on corehq/apps/userreports/pillow.py:
```diff
@@ -5,6 +5,8 @@
 from alembic.autogenerate.api import compare_metadata
 import six
+from sqlalchemy import select
+from sqlalchemy.sql import exists
```
F401 'exists' imported but unused

lwyszomi committed:
flake

emord commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -187,6 +187,30 @@ def test_process_doc_from_sql(self, datetime_mock):
+        datetime_mock.utcnow.return_value = self.fake_time_now
+        sample_doc, expected_indicators = get_sample_doc_and_indicators(self.fake_time_now)
+
+        since = self.pillow.get_change_feed().get_current_offsets()
```
@lwyszomi This recently changed in https://github.com/dimagi/commcare-hq/pull/15420 you'll need to merge master and change this line

lwyszomi replied:
@emord updated

lwyszomi committed:
changes in test regarding PR-15420

lwyszomi committed:
Fix tests for ESAdapter

dimagimon commented on corehq/apps/userreports/pillow.py:
```diff
@@ -5,12 +5,14 @@
 
 from alembic.autogenerate.api import compare_metadata
 import six
+from sqlagg.filters import EQ
```
F401 'EQ' imported but unused

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -218,6 +219,14 @@ def process_change(self, pillow_instance, change):
+                records = query_object.es.doc_id(document['_id'])
+            else:
+                records = table.get_query_object().filter_by(doc_id=document['_id'])
+            return records.count > 0
```
It would be better to add this function to the `table` object directly and then just call `table.doc_exists(doc_id)`

For the SQL table it would be better to use the `exists` function which is more performant than doing a count: http://stackoverflow.com/a/16003360/632517

Commit:
move method to the adapters

dimagimon commented on corehq/apps/userreports/pillow.py:
```diff
@@ -12,6 +12,7 @@
     KAFKA_TOPICS, UCR_ES_BACKEND, UCR_SQL_BACKEND, UCR_LABORATORY_BACKEND
 )
 from corehq.apps.userreports.data_source_providers import DynamicDataSourceProvider, StaticDataSourceProvider
+from corehq.apps.userreports.es.adapter import ESAlchemy
```
F401 'ESAlchemy' imported but unused

dimagimon commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -1,6 +1,8 @@
 from django.utils.translation import ugettext as _
 import sqlalchemy
 from sqlalchemy.exc import IntegrityError, ProgrammingError
+from sqlalchemy.sql.expression import exists
```
F401 'exists' imported but unused

lwyszomi commented:
@snopoke Hey, I moved method to the addapters and now we can user ```table.doc_exists(doc)```, also I changed condition in SQL adapter and now I'm using ```exists``` function.
